extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -360,21 +356,63 @@ export class Bind {
    * Returned promise is resolved when all operations complete.
    * If they don't complete within `timeout` ms, the promise is rejected.
    *
-   * @param {!Array<!Element>} added
-   * @param {!Array<!Element>} removed
+   * @param {!Array<!Element>} addedElements
+   * @param {!Array<!Element>} removedElements
    * @param {number} timeout Timeout in milliseconds.
    * @return {!Promise}
    */
-  scanAndApply(added, removed, timeout = 2000) {
-    dev().info(TAG, 'rescan:', added, removed);
-    const promise = this.removeThenAdd_(removed, added)
-        .then(deltas => {
-          // Don't reevaluate/apply if there are no bindings.
-          if (deltas.added > 0) {
-            return this.evaluate_().then(results =>
-              this.applyElements_(results, added));
-          }
+  scanAndApply(addedElements, removedElements, timeout = 2000) {
+    dev().info(TAG, 'rescan:', addedElements, removedElements);
+    let promise;
+    if (isExperimentOn(this.win_, 'faster-bind-scan')) {
+      /**
+       * Helper function for cleaning up bindings in removed elements.
+       * @param {number} added
+       * @return {!Promise}
+       */
+      const cleanup = added => {
+        this.removeBindingsForNodes_(removedElements).then(removed => {
+          dev().info(TAG,
+              '⤷', 'Δ:', (added - removed), ', ∑:', this.numberOfBindings());
         });
+        return Promise.resolve();
+      };
+      // Early-out if max number of bindings has already been exceeded.
+      // This can happen since we purge old bindings in `removedElements`
+      // _after_ adding new ones (rather than before) as an optimization.
+      if (this.numberOfBindings() > this.maxNumberOfBindings_) {
+        this.emitMaxBindingsExceededError_();
+        return cleanup(0);
+      }
+      const bindings = [];
+      addedElements.forEach(ae => {
+        const elements = ae.querySelectorAll('[i-amphtml-binding]');
+        for (let i = 0; i < elements.length; i++) {
+          this.scanElement_(elements[i], Number.POSITIVE_INFINITY, bindings);
+        }
+      });
+      const added = bindings.length;
+      if (added === 0) {
+        return cleanup(0);
+      }
+      promise = this.sendBindingsToWorker_(bindings).then(() => {
+        return this.evaluate_()
+            .then(results => this.applyElements_(results, addedElements));
+      }).then(() => {
+        // Remove bindings at the end to reduce evaluation/apply latency.
+        // Don't chain this promise since this is a non-blocking clean up task.
```

Comment:
This doesn't really matter, since it's just a `Promise.resolve()`.

Reply:
This is an old comment, thanks.